test/unit/test-sharedmem.cpp

```diff
@@ -0,0 +1,929 @@
+  RAJA::list<RAJA::cuda_shared_mem,
+    RAJA::KernelPolicy<
+      RAJA::statement::CudaKernel<
+        RAJA::statement::For<3, RAJA::cuda_block_exec,
+          RAJA::statement::For<2, RAJA::cuda_block_exec,
+
+          RAJA::statement::CreateShmem<camp::idx_seq<0,1>,
+
+             //Load data into shared memory
+              RAJA::statement::For<1, RAJA::cuda_thread_exec,
+                RAJA::statement::For<0, RAJA::cuda_thread_exec,
+                  RAJA::statement::Lambda<0>
+                                   >
+                                 >,
+              RAJA::statement::CudaSyncThreads,
+
+                //Read data from shared memory
+                RAJA::statement::For<1, RAJA::cuda_thread_exec,
+                  RAJA::statement::For<0, RAJA::cuda_thread_exec,
+                    RAJA::statement::Lambda<1> > >
+              > //close shared memory scope
+            >//for 2
+        >//for 3
+        > //CudaKernel
+      > //kernel policy
+    > //list 
+  >; //types
+INSTANTIATE_TYPED_TEST_CASE_P(CUDA, MatTranspose, CUDATypes);
+#endif
+
+
+template <typename NestedPolicy>
+class MatMultiply : public ::testing::Test
+{
+  virtual void SetUp(){}  
+  virtual void TearDown(){}
+};
+
+TYPED_TEST_CASE_P(MatMultiply);
+
+CUDA_TYPED_TEST_P(MatMultiply, shmem)
+{
+
+  using Tile_pol0 = at_v<TypeParam, 0>;
+  using Tile_size0 = at_v<TypeParam, 1>;
+
+  using Tile_pol1 = at_v<TypeParam, 2>;
+  using Tile_size1 = at_v<TypeParam, 3>;
+  using Pol = at_v<TypeParam, 4>;
+
+  const int DIM = 2;
+
+  //Matrix A size: N x M
+  //Matrix B size: M x P
+  //Result C size: N x P
+
+  const int N = 150;
+  const int M = 25;
+  const int P = 95;
+
+  const int inner_Dim0 = TILE_DIM;
+  const int inner_Dim1 = TILE_DIM;
+
+  const int windowIter = (M-1)/TILE_DIM+1;
+  const int outer_Dim0 = (P-1)/TILE_DIM+1;
+  const int outer_Dim1 = (N-1)/TILE_DIM+1;
+
+  double *A, *B, *C, *C_sol;
+#if defined(RAJA_ENABLE_CUDA)
+  cudaMallocManaged(&A,  sizeof(double) * N * M);
+  cudaMallocManaged(&B,  sizeof(double) * M * P);
+  cudaMallocManaged(&C,  sizeof(double) * N * P);
+  cudaMallocManaged(&C_sol,  sizeof(double) * N * P);
+#else
+  A  = new double[N * M];
+  B  = new double[M * P];
+  C  = new double[N * P];
+  C_sol  = new double[N * P];
+#endif
+
+  RAJA::View<double, RAJA::Layout<DIM>> Aview(A, N, M);
+  RAJA::View<double, RAJA::Layout<DIM>> Bview(B, M, P);
+  RAJA::View<double, RAJA::Layout<DIM>> Cview(C, N, P);
+  RAJA::View<double, RAJA::Layout<DIM>> C_solView(C_sol, N, P);
+
+  for (int row = 0; row < N; ++row) {
+    for (int col = 0; col < M; ++col) {
+      Aview(row, col) = col;
+    }
+  }
+
+  for (int row = 0; row < M; ++row) {
+    for (int col = 0; col < P; ++col) {
+      Bview(row, col) = col;
+    }
+  }
+
+  for(int r=0; r<N; ++r){
+    for(int c=0; c<P; ++c){
+      int dot = 0.0;
+      for(int k=0; k<M; ++k){
+        dot += Aview(r,k)*Bview(k,c);
+      }
+      C_solView(r,c) = dot;
+    }
+  }
+
```

Comment:
@trws , I put together a new iteration of the shared memory framework. Below is a matrix-multiplication example which exercises new RAJA memory objects. In particular, I introduce RAJA memory objects which may be paired with one of the following templates:
RAJA::cpu_tile_mem  - allocates memory on the CPU stack
RAJA::cuda_priv_mem – allocates thread private memory
RAJA::cuda_shared_mem – allocates CUDA shared memory

Furthermore, I introduced a new statement called CreateShmem (name is a work in progress) which enables users to allocate memory in between loops. This enables us to emulate shared memory on the CPU. I was wondering if you might have some time to glance at the work. The code is compatible with a variety of different execution policies as well.